Export selector as byte array instead of u32

cmichi opened the pull request:
Should close https://github.com/paritytech/ink/issues/200.

Needs to be communicated to polkadot-js after approval.

Robbepop commented:
> I've encoded with a prefix (`"0xF0…"` instead of `"F0…"`), since I think it makes the string self explanatory whereas the other representation isn't (and it's just two bytes more). Just ping me if you think there are reasons against this.

This is exactly what I would have done. 👍

Robbepop commented on abi/src/layout.rs:
```diff
@@ -205,3 +209,17 @@ impl LayoutRange {
+    }
+
+    serializer.serialize_str(&hex)
+}
```
Could you provide a simple test that shows the result of this encoding as string or JSON for some given keys?

Robbepop commented on abi/src/specs.rs:
```diff
@@ -190,6 +190,7 @@ pub struct ConstructorSpec<F: Form = MetaForm> {
     name: F::String,
     /// The selector hash of the message.
+    #[serde(serialize_with = "serialize_selector")]
     selector: u32,
```
This solution is not bad since it is kept pretty local. However, what we really want is to have `[u8; 4]` at this position fed by `ink_lang` correctly. With this PR we would have the situation of `ink_lang` converting its `[u8; 4]` magically into a `u32` just so that we can finally convert it back into `[u8; 4]` before serializing. This is also dangerous since nothing guarantees us from `ink_lang` how the conversion from `[u8; 4]` actually works.

cmichi replied:
Ok, so to make sure I understand you correctly: What you suggest is to instead replace the `ink_lang` internal `selector` type `u32` with `[u8; 4]` everywhere, correct?

Robbepop replied:
Yes but this is not a local change as your current PR is. So I can totally understand why you went this way. However, this double transformation that we have with this PR is definitely not something I want to elaborate further. So let us put in a bit more work and fix it properly. :)

Robbepop commented on abi/src/specs.rs:
```diff
@@ -795,3 +797,12 @@ impl MessageParamSpecBuilder {
+    let b = selector.to_be_bytes();
+    let hex = format!("0x{:02X}{:02X}{:02X}{:02X}", b[0], b[1], b[2], b[3]);
+    serializer.serialize_str(&hex)
+}
```
Please also provide a simple test case for this conversion. (Will be super simple since we are going to start from `[u8; 4]` according to the other review.

cmichi commented:
I updated the PR with all comments addressed. I added some comments to the macro, not sure how Rust-idiomatic this is, since I only see comments in macros seldomly.

Robbepop commented on lang/src/gen/abi.rs:
```diff
@@ -137,6 +138,13 @@ fn generate_abi_messages<'a>(
+        );
+        let selector = parse_str::<syn::Expr>(&selector)
+            .expect("failed to parse selector");
+
```
These transformations are required because `syn` doesn't provide a `ToTokens` impl for `[u8; 4]`, right?

cmichi replied:
Yup, right! I searched for a more elegant solution, but haven't found one yet.

Robbepop commented on lang/src/gen/as_dependency.rs:
```diff
@@ -237,6 +238,11 @@ fn generate_call_enhancer_messages<'a>(
+                "[{}, {}, {}, {}]",
+                selector[0], selector[1], selector[2], selector[3]
+            );
+            let selector = parse_str::<syn::Expr>(&selector).expect("failed to parse selector");
```
same question here. i see a potential code dupe. maybe we want to pull that code out and do this just once.

Robbepop commented on lang/src/gen/build.rs:
```diff
@@ -474,11 +474,12 @@ fn codegen_for_messages(tokens: &mut TokenStream2, contract: &hir::Contract) {
+                msg_selector[0], msg_selector[1], msg_selector[2], msg_selector[3]
             );
+            let msg_id = parse_str::<syn::Expr>(&msg_selector)
+                .expect("failed to parse msg_id");
```
Third, so please do a function and prevent code dupe here. :)

Robbepop commented on model/src/msg.rs:
```diff
@@ -40,44 +40,122 @@ pub trait Message {
+    // unambiguously if it should match the doc comment to `meta` or
+    // `expr`. Hence we introduce a `@delimiter` between the two of them.
+    // The first two rules match for macro invocations containing this
+    // delimiter, the subsequent ones add the delimiter if it's missing.
```
Thank you for this comment! :+1:

liuchengxu commented:
I'm not sure if this commit is actually related. But I failed to compile the examples/lang/erc20 at this commit:

```bash
$ cd ink
$ cargo build
$ cd examples/lang/erc20
$ ../../../target/debug/cargo-contract contract build
 [1/3] Collecting crate metadata
 [2/3] Building cargo project
       Fresh unicode-xid v0.2.0
       Fresh unicode-xid v0.1.0
       Fresh lazy_static v1.4.0
       Fresh regex-syntax v0.6.12
       Fresh cfg-if v0.1.10
       Fresh memory_units v0.4.0
       Fresh nodrop v0.1.14
       Fresh smallvec v0.6.10
       Fresh either v1.5.3
       Fresh itoa v0.4.4
       Fresh unicode-segmentation v1.3.0
       Fresh thread_local v0.3.6
       Fresh itertools v0.8.0
       Fresh heck v0.3.1
       Fresh proc-macro2 v1.0.6
       Fresh proc-macro2 v0.4.30
       Fresh memchr v2.2.1
       Fresh libc v0.2.65
       Fresh crunchy v0.2.2
       Fresh arrayvec v0.4.12
       Fresh ryu v1.0.2
       Fresh quote v1.0.2
       Fresh quote v0.6.13
       Fresh aho-corasick v0.7.6
       Fresh wee_alloc v0.4.5
       Fresh tiny-keccak v1.5.0
       Fresh syn v1.0.5
       Fresh syn v0.15.44
       Fresh regex v1.3.1
       Fresh ink_alloc v0.1.0 (/home/xlc/src/github.com/paritytech/ink/alloc)
       Fresh serde_derive v1.0.101
       Fresh derive_more v0.15.0
       Fresh ink_utils v0.1.0 (/home/xlc/src/github.com/paritytech/ink/utils)
       Fresh serde v1.0.101
       Fresh toml v0.5.3
       Fresh serde_json v1.0.41
       Fresh proc-macro-crate v0.1.4
       Fresh parity-scale-codec-derive v1.0.3
       Fresh parity-scale-codec v1.0.6
       Fresh ink_core v0.1.0 (/home/xlc/src/github.com/paritytech/ink/core)
       Fresh ink_model v0.1.0 (/home/xlc/src/github.com/paritytech/ink/model)
       Fresh ink_lang v0.1.0 (/home/xlc/src/github.com/paritytech/ink/lang)
   Compiling erc20 v0.1.0 (/home/xlc/src/github.com/paritytech/ink/examples/lang/erc20)
     Running `rustc --edition=2018 --crate-name erc20 src/lib.rs --color never --crate-type cdylib --crate-type rlib --emit=dep-info,link -C opt-level=z -C panic=abort -C overflow-checks=on -C metadata=b380712fdb95c74a --out-dir /home/xlc/src/github.com/paritytech/ink/examples/lang/erc20/target/wasm32-unknown-unknown/release/deps --target wasm32-unknown-unknown -L dependency=/home/xlc/src/github.com/paritytech/ink/examples/lang/erc20/target/wasm32-unknown-unknown/release/deps -L dependency=/home/xlc/src/github.com/paritytech/ink/examples/lang/erc20/target/release/deps --extern ink_core=/home/xlc/src/github.com/paritytech/ink/examples/lang/erc20/target/wasm32-unknown-unknown/release/deps/libink_core-fc3986115ca64ed4.rlib --extern ink_lang=/home/xlc/src/github.com/paritytech/ink/examples/lang/erc20/target/release/deps/libink_lang-25234a970a87aa94.so --extern ink_model=/home/xlc/src/github.com/paritytech/ink/examples/lang/erc20/target/wasm32-unknown-unknown/release/deps/libink_model-a4568562f0acd6ad.rlib --extern scale=/home/xlc/src/github.com/paritytech/ink/examples/lang/erc20/target/wasm32-unknown-unknown/release/deps/libparity_scale_codec-e53ab60f102d9285.rlib -C 'link-args=-z stack-size=65536 --import-memory'`
error: local ambiguity: multiple parsing options: built-in NTs expr ('msg_id') or 1 other option.
   --> src/lib.rs:101:9
    |
101 |         /// on the behalf of the message's sender.
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to previous error

error: could not compile `erc20`.

Caused by:
  process didn't exit successfully: `rustc --edition=2018 --crate-name erc20 src/lib.rs --color never --crate-type cdylib --crate-type rlib --emit=dep-info,link -C opt-level=z -C panic=abort -C overflow-checks=on -C metadata=b380712fdb95c74a --out-dir /home/xlc/src/github.com/paritytech/ink/examples/lang/erc20/target/wasm32-unknown-unknown/release/deps --target wasm32-unknown-unknown -L dependency=/home/xlc/src/github.com/paritytech/ink/examples/lang/erc20/target/wasm32-unknown-unknown/release/deps -L dependency=/home/xlc/src/github.com/paritytech/ink/examples/lang/erc20/target/release/deps --extern ink_core=/home/xlc/src/github.com/paritytech/ink/examples/lang/erc20/target/wasm32-unknown-unknown/release/deps/libink_core-fc3986115ca64ed4.rlib --extern ink_lang=/home/xlc/src/github.com/paritytech/ink/examples/lang/erc20/target/release/deps/libink_lang-25234a970a87aa94.so --extern ink_model=/home/xlc/src/github.com/paritytech/ink/examples/lang/erc20/target/wasm32-unknown-unknown/release/deps/libink_model-a4568562f0acd6ad.rlib --extern scale=/home/xlc/src/github.com/paritytech/ink/examples/lang/erc20/target/wasm32-unknown-unknown/release/deps/libparity_scale_codec-e53ab60f102d9285.rlib -C 'link-args=-z stack-size=65536 --import-memory'` (exit code: 1)
error: BuildFailed
```

@cmichi  Any clue?

cmichi commented:
@liuchengxu Thanks for the comment! Yes, it's indeed an issue with the macro which appears when there are doc comments over multiple lines. I'm looking into fixing it.

cmichi commented:
@liuchengxu You should be able to build the contract now on `master`.
